product_contract/models/sale_order.py

```diff
@@ -0,0 +1,142 @@
+                    and r
+                    not in r.contract_id.contract_line_ids.mapped("sale_order_line_id")
+                )
+                if line_to_create_contract or line_to_update_contract:
+                    rec.need_contract_creation = True
+
+    @api.depends("order_line")
+    def _compute_is_contract(self):
+        self.is_contract = any(self.order_line.mapped("is_contract"))
+
+    def _prepare_contract_value(self, contract_template):
+        self.ensure_one()
+        return {
+            "name": "{template_name}: {sale_name}".format(
+                template_name=contract_template.name, sale_name=self.name
+            ),
+            "partner_id": self.partner_id.id,
+            "company_id": self.company_id.id,
+            "contract_template_id": contract_template.id,
+            "user_id": self.user_id.id,
+            "payment_term_id": self.payment_term_id.id,
+            "fiscal_position_id": self.fiscal_position_id.id,
+            "invoice_partner_id": self.partner_invoice_id.id,
+            "line_recurrence": self.partner_invoice_id.id,
+        }
+
+    def action_create_contract(self):
+        contract_model = self.env["contract.contract"]
+        contracts = self.env["contract.contract"]
+        for rec in self.filtered("is_contract"):
+            line_to_create_contract = rec.order_line.filtered(
+                lambda r: not r.contract_id and r.product_id.is_contract
+            )
+            line_to_update_contract = rec.order_line.filtered(
+                lambda r: r.contract_id
+                and r.product_id.is_contract
+                and r
+                not in r.contract_id.contract_line_ids.mapped("sale_order_line_id")
+            )
+            contract_templates = self.env["contract.template"]
+            for order_line in line_to_create_contract:
+                contract_template = order_line.product_id.with_company(
+                    rec.company_id
+                ).property_contract_template_id
+                if not contract_template:
+                    raise ValidationError(
+                        _(
+                            "You must specify a contract "
+                            "template for '{}' product in '{}' company."
+                        ).format(order_line.product_id.name, rec.company_id.name)
+                    )
+                contract_templates |= contract_template
+            for contract_template in contract_templates:
+                order_lines = line_to_create_contract.filtered(
+                    lambda r, template=contract_template: r.product_id.with_company(
+                        r.order_id.company_id
+                    ).property_contract_template_id
+                    == template
+                )
+                contract = contract_model.create(
+                    rec._prepare_contract_value(contract_template)
+                )
+                contracts |= contract
```

Comment:
Better in terms of performances to build a set of ids and then return at the end : contract_model.browse(ids)